Subscription plan

lBenevides opened the pull request:
- Criação do model Subscription
- Criação da index
- Criação do show
- Criação do new e create
- Testes unitarios feitos
- Testes de rotas criados
- Autenticação restrita para admin para o new e create

Matheustlacerda committed:
teste de index concluído

Co-authored-by: Lucas <lhenriquebenevides@gmail.com>

Matheustlacerda committed:
testes criar e ver planos de assinatura

Co-authored-by: Lucas <lhenriquebenevides@gmail.com>

lBenevides committed:
Testes de systemas feitos e refatorados,testes de models feito e  Fabricate criado para Admin/User/Subscription

lBenevides committed:
Testes de rota feitos

lBenevides committed:
Adicionado index unique para coluna name, tabela subscription

lBenevides commented:
TODO: 
Adicionar os cursos ao plano de assinatura, fazer um CRUD disso. Não sei se a melhor maneira seria por tags, colocando uma array de tags no banco ou adicionando cada curso individualmente, por meio de uma tabela com associação. 

Acho que com a tabela por associação ficaria melhor

lBenevides committed:
Admin pode adicionar um curso a uma assinatura

lBenevides committed:
rubocup ajustado

HenriqueMorato commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "liveServer.settings.port": 5501
+}
```
podemos colocar essa pasta no gitignore?

lBenevides replied:
claro, acabei deixando isso passar sem nem ver

HenriqueMorato replied:
Acho que pra esse aqui sumir você vai ter que fazer `git rm -r --cached .vscode`

HenriqueMorato commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,46 @@
+class SubscriptionsController < ApplicationController
+  before_action :authenticate_admin!, only: %i[new create add_course]
```
O index e o show não vai precisar?

lBenevides replied:
Por enquanto não, um usuário ou visitante poderia ter acesso a essas views

HenriqueMorato commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  def add_course
+    @course = Course.find_by(name: params[:name])
+
+    if @course.present?
```
Como essa validação é usada na navegação?

lBenevides replied:
A validação é para verificar se o curso foi encontrado ou não no find_by. Poderia ficar só **if @course** certo?

HenriqueMorato replied:
Assim fica melhor se for isso, eu sugeri um esquema um pouco diferente pra não dar tanto trabalho

HenriqueMorato commented on app/views/subscriptions/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<hr>
+<% @subscriptions.each do |subscription| %>
+  <div id="<%= subscription.id %>">
+    <%= link_to subscription.name, subscription_path(subscription) %><br>
```
```suggestion
    <%= link_to subscription.name, subscription %><br>
```

HenriqueMorato commented on app/views/subscriptions/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<%= link_to 'Criar novo plano de assinatura', new_subscription_path %>
+<%= flash[:alert] %>
```
Esse alert já existe no layout acho que não é necessário.

lBenevides replied:
beleza

HenriqueMorato commented on app/views/subscriptions/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+  <dd><%= @subscription.price %></dd>
+</dl>
+
+<%= form_with url: add_course_subscription_path do |f| %>
+  <%= f.label :name, 'Nome do curso' %> <br>
+  <%= f.text_field :name %> <br>
+  <%= f.submit 'Adicionar curso' %> 
+<% end %>
```
O que acham de colocar um botão para cada curso que não está nessa assinatura? A busca por nome é muito sujeita a falhas

HenriqueMorato replied:
Talvez pudesse usar um select também ou coisa assim

lBenevides replied:
Coloquei como nome só para fazer funcionar, mas a ideia é que isso seja uma busca por nome, dai apareceria uma lista de cursos e poderia escolher quais adicionar.

lBenevides replied:
Refatorado! por meio do nome da para buscar os cursos e escolher a opção de adicionar cada um

HenriqueMorato commented on app/views/subscriptions/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+<% end %>
+<hr>
+
+<% if @subscription.courses.any? %>
```
Como não tem nada customizado pra essa tela e nem um `else` desse `any?` ele não parece necessário a principio.

O que acham colocar um titulo nesse local para mostrar o que são esses cursos.

lBenevides replied:
Faltou o else com 'nenhum curso cadastrado nesse plano de assinatura', vou ajustar aqui e colocar um titulo

HenriqueMorato commented on spec/models/subscription_course_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe SubscriptionCourse, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Como esse arquivo não está sendo usado podemos removê-lo?

lBenevides replied:
removido!

HenriqueMorato commented on spec/models/subscription_spec.rb:
```diff
@@ -1,5 +1,20 @@
+  it 'attributes cannot be in blank' do
+    subscription = Subscription.new
+
+    refute subscription.valid?
```
Essa sintaxe não é muito usada no rspec, o que acham de:

```suggestion
    expect(subscription).not_to be_valid
```

lBenevides replied:
Tentei assim e acabou dando crash

Subscription name should be uniq
     Failure/Error: expect(subscription).not_to be_valid?
       expected #<Subscription id: nil, name: "Jornada Web com Rails", description: nil, created_at: nil, updated_at: nil, price: nil, token: nil> to respond to `valid??`
     # ./spec/models/subscription_spec.rb:17:in `block (2 levels) in <top (required)>'

lBenevides replied:
feito!

HenriqueMorato replied:
Acho que essa não foi não

HenriqueMorato commented on spec/request/subscription_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    login_as user, scope: :user
+
+    post subscriptions_path, params: {
+      subscription: { name: 'Teste', description: 'Teste teste', price: 'teste' }
```
Aqui daria pra usar os atributos do fabrication se quiserem

lBenevides replied:
como ficaria? Fabricate(:subscription).name ?

HenriqueMorato replied:
Usa aquele lance dos atributos num hash que o Fabrication tem

lBenevides replied:
ah, so passar o fabricate(:subscription) como so argumentos, feito

post subscriptions_path, params: {
      subscription: Fabricate(:subscription)
    }

HenriqueMorato replied:
Idealmente seria Fabricate.attributes_for(:subscription) pra receber só um hash

params { Fabricate.attributes_for(:subscription) }

HenriqueMorato commented on spec/system/subscription_spec.rb:
```diff
@@ -0,0 +1,108 @@
+
+describe 'Subscription plan' do
+  it 'view all subscriptions plans' do
+    Subscription.create!(name: 'Jornada Web com Rails',
+                         description: 'Esta assinatura irá englobar todos os cursos de Ruby e Rails',
+                         price: '50')
+    Subscription.create!(name: 'Jornada Mobile com Dart e Flutter',
+                         description: 'Esta assinatura irá englobar todos os cursos de Dart e o framework Flutter',
+                         price: '80')
```
O que acham de usar Fabrication?

lBenevides replied:
feito! este teste foi o primeiro que fiz, n tinha criado os Fabricator na hora e esqueci de refatorar

HenriqueMorato commented on spec/system/subscription_spec.rb:
```diff
@@ -0,0 +1,108 @@
+                         price: '80')
+
+    visit subscriptions_path
+    # TODO: colocar um within aqui, ta dando bug por alguma razao
```
deu certo essa aqui?

lBenevides replied:
Não funcionou, tava tentando fazer um 'div#id' mas não funcinou de jeito maneira. 

Sempre da esse erro:
 Failure/Error:
       within 'div#1' do
         expect(page).to have_text(subscription.name)
         expect(page).to have_text(subscription.description)
       end
     
     Selenium::WebDriver::Error::InvalidSelectorError:
       invalid selector: An invalid or illegal selector was specified
         (Session info: headless chrome=90.0.4430.72)

HenriqueMorato replied:
Acho que é porque o id só é numérico coloca um texto antes que talvez dê certo.

lBenevides replied:
Não entendi, como assim?

HenriqueMorato replied:
within 'div#1' você estaria chamando o ID e 1 é isso mesmo?

lBenevides replied:
Entendi! Só numerico não funciona, tive que colocar algum nome antes do numero

lBenevides committed:
Update app/views/subscriptions/index.html.erb

Co-authored-by: Henrique Morato <riquemorato@gmail.com>

lBenevides committed:
Codigo com ajustes pos review de henrique

lBenevides committed:
commits feitos no review
Merge branch 'Subscription_plan' of github.com:iugu-for-devs/codeplay into Subscription_plan

lBenevides committed:
Adicionar curso refatorado

JorgeLAB commented on db/migrate/20210414202445_add_price_to_subscription.rb:
```diff
@@ -0,0 +1,5 @@
+class AddPriceToSubscription < ActiveRecord::Migration[6.1]
+  def change
+    add_column :subscriptions, :price, :decimal, null: false
```
Seria interessante também configurar a prescion e scale para ter um retorno definido:
`t.decimal :price, precision: 8, scale: 2`

JorgeLAB commented on spec/fabricators/subscription_course_fabricator.rb:
```diff
@@ -0,0 +1,4 @@
+Fabricator(:subscription_course) do
+  course       nil
```
Acho que daria para tirar os `nils` o/

JorgeLAB commented on app/views/subscriptions/_course.html.erb:
```diff
@@ -0,0 +1,18 @@
+      <p><%= course.name %></p>  
+      <p><%= course.description %></p>
+
+      <%= form_with url: add_course_subscription_path do |form| %>
```
Aqui acho que daria para fazer só um `<%= link_to "Adicionar Curso", add_course_subscripition_path( passando_course ), method: post %>`
=> https://stackoverflow.com/questions/13414663/using-rails-link-to-for-links-that-post  algo assim em um link.

lBenevides replied:
da não, é preciso passar o parametro do curso que vai ser adicionado... fiz uma gambiarra com hidden form, mas isso da pra ser ajustado de outra maneira

HenriqueMorato replied:
Pra funcionar esse jeito via link você teria que mudar suas rotas pra algo tipo:

```
resources :subscriptions do
  resources :courses, only: [] do
    post :add, on: :member
  end
end
```

que aí o link ficaria algo como `add_course_subscription_path(@subscription, @course)` mas acho que você podia tentar algo assim. Trocar todo esse each para:

```
<%= form_with url: add_course_subscription_path(@subscription) do |form| %>
  <%= f.collection_select :course_id, @courses, :id, :name %>
  <%= f.submit 'Adicionar curso' %>
<% end %>
```

Porque aí esse curso ai invés de ir como um hidden field você manda o ID dele através de um select. Daria pra fazer via botão também mas ficaria meio estranho assim:

<%= link_to "Adicionar Curso", add_course_subscription_path(@subscription, course_id: course.id), method: post %>

JorgeLAB commented on app/views/subscriptions/new.html.erb:
```diff
@@ -0,0 +1,26 @@
+<%= link_to 'Voltar', subscriptions_path %>
+
+<% if @subscription.errors.any?%>
```
O pessoal da campus colocou flash configurado para renderizar em qualquer view. Está na pasta `shared` não acredito que tenha necessidade de colocá-lo aqui o flash deles já faz essa iteração de erros.

lBenevides replied:
?

JorgeLAB replied:
Tipo, essa parte toda de erro que você colou poderia ser removida:
```
<% if @subscription.errors.any?%>
  <h4>Não foi possivel salvar. Encontramos alguns erros:</h4>
  <ul>
  <% @subscription.errors.full_messages.each do |error|  %>
    <li><%= error %></li>
  <% end %>
  </ul>
<% end %>
```
Lá no controller do @subscription você poderia colocar:
````
    if @subscription.save
      redirect_to @subscription, alert: 'Plano cadastrado com sucesso'
    else
      flash.now[:notice] = @subscription.errors.full_messages
      render :new
    end
```
O `flash.now` o/

lBenevides replied:
ahhh saquei

lBenevides committed:
Conflitos de merge resolvidos

HenriqueMorato commented on .gitignore:
```diff
@@ -40,3 +40,5 @@ yarn-debug.log*
 
 coverage
+# Ignore vscode settings
+/.vscode/settings.json
```
Pode tirar a pasta toda vai que surge mais arquivos 🤣 

```suggestion
.vscode
```

lBenevides replied:
beleza haha

lBenevides committed:
Ajustes em codigos para merge

lBenevides committed:
Conflitos resolvidos

HenriqueMorato commented on config/routes.rb:
```diff
@@ -2,6 +2,13 @@
 
+  resources :subscriptions, only: %i[index new create show] do
+    get 'search_course', on: :member
+    resources :courses, only: [] do
+      post 'add', on: :member
+    end
```
A rota acho que pode ficar como era antes um add_course on member. Essa aqui era só se você quisesse usar um link

HenriqueMorato commented on app/views/subscriptions/show.html.erb:
```diff
@@ -0,0 +1,42 @@
+<% end %>
+
+
+<% if !@courses.nil? %>
```
Acho que essa nunca vai acontecer né? porque na teoria o mínimo é um array vazio, ou não?

HenriqueMorato reviewed (approved):
Vou deixar pré aprovado
